stdenv, cacert: $NIX_SSL_CERT_FILE changes

vcunat opened the pull request:
###### Motivation for this change

Some SSL libs don't react to `$SSL_CERT_FILE`.  That actually makes sense to me, as we add this behavior as nixpkgs-specific, so it seems "safer" to use `$NIX_*`.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date: _docs are **still** missing ;-)_
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

vcunat committed:
stdenv, cacert: consider $NIX_SSL_CERT_FILE in hooks

Some SSL libs don't react to $SSL_CERT_FILE.
That actually makes sense to me, as we add this behavior
as nixpkgs-specific, so it seems "safer" to use $NIX_*.

layus commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -646,9 +646,12 @@ fi
+if [ -z "${NIX_SSL_CERT_FILE:-}" ] && [ -z "${IN_NIX_SHELL:-}" ]; then
+  export NIX_SSL_CERT_FILE=/no-cert-file.crt
+fi
+# Another variant left for compatibility.
```
I think it would make sense to enable this in pure shells, for consistency. But that would be for another PR.

vcunat replied:
:heart:  see the new commit.

layus reviewed (approved):
It totally makes sense, and should have been done before :-)

vcunat committed:
stdenv: also override cert files in pure nix-shell

That's very much consistent with the spirit of nix-shell --pure

BTW, nix 1.x shells will be always treated as pure;
in that version detection isn't possible.
https://github.com/NixOS/nix/commit/1bffd83e1a9c

vcunat committed:
pypi2nix: fix $IN_NIX_SHELL test

In nix 2.0 this changed: https://github.com/NixOS/nix/commit/1bffd83
I only kept the original intention and did no kind of verification.

vcunat commented:
/cc pypi2nix maintainer @garbas.  It couldn't work for years in the intended way, so I don't know.

LnL7 commented:
My reasoning back when I added it was to go trough the upstream codepath instead of our patches by default. But that doesn't really make sense since we want NIX_SSL_CERT_FILE to work for everything.
